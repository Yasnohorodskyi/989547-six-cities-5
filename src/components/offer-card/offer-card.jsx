import React from "react";
import {Link} from "react-router-dom";
import propTypes from "prop-types";
import {connect} from "react-redux";
import {hoverOffer} from "../../store/action.js";

const adaptToClient = (data) => {
  const adaptedData = Object.assign(
      {},
      data,
      {
        isPremium: data.is_premium,
        previewImg: data.preview_image,
        cost: data.price,
        isFavorite: data.is_favorite,
        rate: data.rating
      }
  );

  delete adaptedData.is_premium;
  delete adaptedData.preview_image;
  delete adaptedData.price;
  delete adaptedData.is_favorite;
  delete adaptedData.rating;

  return adaptedData;
};

const OfferCard = ({onHoverOfferAction, className, block, offer}) => {
  const {id, isPremium, previewImg, cost, isFavorite, rate, title, type} = adaptToClient(offer);

  const onMouseOver = () => onHoverOfferAction(offer);
  return (
    <article className={`${className} place-card`} onMouseOver={onMouseOver}>
      {
        isPremium
          ? <div className="place-card__mark">
            <span>Premium</span>
          </div>
          : ``
      }
      <div className={`${block}__image-wrapper place-card__image-wrapper`}>
        <Link to={{
          pathname: `/offer/${id}`,
          offer
        }}>
          <img className="place-card__image" src={previewImg} width="260" height="200"/>
        </Link>
      </div>
      <div className="place-card__info">
        <div className="place-card__price-wrapper">
          <div className="place-card__price">
            <b className="place-card__price-value">&euro;{cost}</b>
            <span className="place-card__price-text">&#47;&nbsp;night</span>
          </div>
          <button className={
            isFavorite
              ? `place-card__bookmark-button place-card__bookmark-button--active button`
              : `place-card__bookmark-button button`
          } type="button">
            <svg className="place-card__bookmark-icon" width="18" height="19">
              <use xlinkHref="#icon-bookmark"></use>
            </svg>
            <span className="visually-hidden">To bookmarks</span>
          </button>
        </div>
        <div className="place-card__rating rating">
          <div className="place-card__stars rating__stars">
            <span style={{width: `${rate * 20}%`}}></span>
            <span className="visually-hidden">Rating</span>
          </div>
        </div>
        <h2 className="place-card__name">
          <Link to={{
            pathname: `/offer/${id}`,
            offer
          }}>{title}</Link>
        </h2>
        <p className="place-card__type">{type}</p>
      </div>
    </article>
  );
};

const mapDispatchToProps = (dispatch) => ({
  onHoverOfferAction(offer) {
    dispatch(hoverOffer(offer));
  }
});

OfferCard.propTypes = {
  offer: propTypes.object,
  className: propTypes.string.isRequired,
  block: propTypes.string.isRequired,
  onHoverOfferAction: propTypes.func,
};

export {OfferCard};
export default connect(``, mapDispatchToProps)(OfferCard);
