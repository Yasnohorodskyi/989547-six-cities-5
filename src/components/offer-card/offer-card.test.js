import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";
import configureMockStore from "redux-mock-store";

import {OfferCard} from "./offer-card";

const offer = {
  id: 123,
  city: {
    name: `Amsterdam`
  },
  cost: 444,
  isFavorite: true,
  isPremium: false,
  previewImg: `https://picsum.photos/210`,
  rate: 4,
  title: `Bla bla bla`,
  type: `Room`
};

const noop = () => {};
const className = `cities__place-card`;
const block = `cities`;
const imgWidth = `123`;
const imgHeight = `56`;

const mockStore = configureMockStore()({
  PROCESS: {
    hoveredOffer: {}
  }
});

it(`Should Offer Card render`, () => {
  const tree = renderer
  .create(
      <Provider store={mockStore}>
        <MemoryRouter>
          <OfferCard
            offer={offer}
            className={className}
            block={block}
            onHoverOfferAction={noop}
            changeBookmarkStatus={noop}
            imgWidth={imgWidth}
            imgHeight={imgHeight}
          />
        </MemoryRouter>
      </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
