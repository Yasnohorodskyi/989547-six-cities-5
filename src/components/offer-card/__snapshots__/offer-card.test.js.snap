// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Offer Card render 1`] = `
<article
  className="cities__place-card place-card"
  onMouseOver={[Function]}
>
  
  <div
    className="cities__image-wrapper place-card__image-wrapper"
  >
    <a
      href="/offer/123"
      onClick={[Function]}
    >
      <img
        className="place-card__image"
        height="56"
        src="https://picsum.photos/210"
        width="123"
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          444
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <button
        className="place-card__bookmark-button button place-card__bookmark-button--active"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="place-card__bookmark-icon"
          height={19}
          width={18}
        >
          <use
            xlinkHref="#icon-bookmark"
          />
        </svg>
        <span
          className="visually-hidden"
        >
          To bookmarks
        </span>
      </button>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "80%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        href="/offer/123"
        onClick={[Function]}
      >
        Bla bla bla
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      Room
    </p>
  </div>
</article>
`;
